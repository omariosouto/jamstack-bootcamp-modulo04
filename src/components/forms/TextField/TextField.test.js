import React from 'react';
import user from '@testing-library/user-event';
import TextField from './index';
import { render, screen } from '../../../infra/test/testUtils';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        name="name"
        value="Mario"
        onChange={() => {}}
      />,
    );

    // screen.debug();

    expect(screen.getByPlaceholderText('Nome')).toHaveValue('Mario');
    expect(screen.getByPlaceholderText('Nome')).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const { container } = render(
          <TextField
            placeholder="Email"
            name="email"
            onChange={() => {}}
            value="teste_teste.com"
            isTouched
            error={{
              message: 'You must provide a valid email',
            }}
          />,
        );

        expect(screen.getByPlaceholderText('Email')).toHaveValue('teste_teste.com');
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          placeholder="Email"
          name="email"
          onChange={() => {}}
          value="teste_teste.com"
          isTouched
          error={{
            message: 'You must provide a valid email',
          }}
        />,
      );

      expect(screen.getByPlaceholderText('Email')).toHaveValue('teste_teste.com');
      expect(screen.getByRole('alert')).toHaveTextContent('You must provide a valid email');
      expect(screen.getByPlaceholderText('Email')).toMatchSnapshot();
    });
  });

  describe('when user types', () => {
    test('calls onChange handler', async () => {
      const onChange = jest.fn();
      render(
        <TextField
          placeholder="Email"
          name="email"
          onChange={onChange}
          value=""
        />,
      );

      const inputEmail = screen.getByPlaceholderText('Email');
      user.type(inputEmail, 'mario@gmail.com');

      expect(onChange).toHaveBeenCalledTimes(15);
    });
  });
});
