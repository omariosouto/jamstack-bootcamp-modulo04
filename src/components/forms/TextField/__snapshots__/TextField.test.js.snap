// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 8px;
}

<input
  class="c0 c1"
  name="name"
  placeholder="Nome"
  type="text"
  value="Mario"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 8px;
  border-color: #dc3545;
}

.c1 + span {
  color: #dc3545;
  font-size: 11px;
}

<input
  class="c0 c1"
  name="email"
  placeholder="Email"
  type="text"
  value="teste_teste.com"
/>
`;

exports[`<TextField /> when field is valid and user is typing the value must be updated 1`] = `
.c1 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c3 {
  font-size: 12px;
  font-weight: 400;
  line-height: 1;
  color: #dc3545;
}

.c0 {
  margin-bottom: 17px;
}

.c2 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 8px;
  border-color: #dc3545;
}

.c2 + span {
  color: #dc3545;
  font-size: 11px;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1 c2"
      name="email"
      placeholder="Email"
      type="text"
      value="teste_teste.com"
    />
    <span
      class="c3"
      color="error.main"
      id="textField_error_label_email"
      role="alert"
    >
      You must provide a valid email
    </span>
  </div>
</div>
`;
